load esper queries from db

GeorgeJahad opened the pull request:
# Resolves

https://jira.rax.io/browse/SALUS-1038
https://jira.rax.io/browse/SALUS-1039


# What

Implement metric state calculation logic in event processor.  

Note that this PR is against the jjbuchan/new-consumer branch, not master.

# How

[design](https://github.com/Rackspace-Segment-Support/salus-docs-internal/blob/master/proposals/event-engine/queryLoad.md)

## How to test

run the unit tests


# TODO
1. A number of the UniversalMetricHandler tests seemed to be failing intermittently even before I merged my changes.  We'll need to investigate those.

2. Haven't implemented the "rate" custom metric.  Will create separate PR for that.

3. Haven't made these [additions/removals](https://github.com/Rackspace-Segment-Support/salus-docs-internal/blob/master/proposals/event-engine/queryLoad.md#fields-to-be-modified).  Will create separate PR for that.

4. Not forwarding custom metrics back to the metrics topic. Will create separate PR for that.


# Questions

1. Should the event engine really whether the  monitorScope is remote or local?

2. Should this PR use the [message field in state expression](https://github.com/racker/salus-telemetry-model/blob/58b4639234e74b32fe008fb64c8a00ca7c158904/src/main/java/com/rackspace/salus/telemetry/entities/EventEngineTaskParameters.java#L100)?  It is not currently.

GeorgeJahad commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+    }
+
+    // Check each type of stateExpression and return the first one that is true
+    for (String state : List.of("CRITICAL", "WARNING", "OK")) {
```
This logic seems a bit goofy to me.  If the critical and warning stateExpressions are false, it then evaluates the ok expression, but below it sets it to ok in any case, so the ok stateExpression is actually redundant.  I'm probably going to remove it.  What do you guys think?

GeorgeJahad replied:
I'm fine with it as long as it has some other side effect like setting the message differently, (or even just a different log message.)  but if the results are indistinguishable, i wouldn't waste time doing it.

itzg commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+    return taskDataMap.remove(taskId).getDeploymentId();
+  }
+
+  public static SalusEnrichedMetric initMetric(SalusEnrichedMetric metric, String state,
```
The naming of this method and `generateEnrichedMetric` confuse me a little since both seem to be doing more updating rather than initializing/creating. Maybe javadoc on each would help.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+    // With Operator.AND return as soon as false is found
+    if (expression.getOperator() == Operator.AND) {
+      for (Expression subExpression : expression.getExpressions()) {
+        if (subExpression instanceof LogicalExpression) {
+          retval = getLogicalValue((LogicalExpression) subExpression, metric, customMetricData);
+
+        } else {
+          retval = getComparisonValue((ComparisonExpression) subExpression,
+              metric, customMetricData);
+        }
```
```suggestion
        retval = getStateExpressionValue(subExpression, metric, customMetricData);
```

Can something like this not be used instead?

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+      return true;  // Operator.AND is true if all are true
+    } else {      // With Operator.OR, return as soon as true is found
+      for (Expression subExpression : expression.getExpressions()) {
+        if (subExpression instanceof LogicalExpression) {
+          retval = getLogicalValue((LogicalExpression) subExpression, metric, customMetricData);
+
+        } else {
+          retval = getComparisonValue((ComparisonExpression) subExpression,
+              metric, customMetricData);
+        }
```
```suggestion
        retval = getStateExpressionValue(subExpression, metric, customMetricData);
```

Same again here

itzg commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+    return stringCompareFunctions.get(expression.getComparator()).apply(operand1, operand2);
+  }
+
+  @Data
```
On classes that have business-y logic like `eval`, it feels better not use to `@Data` on them. Instead, declare the specific accessor/constructors that are needed.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+  private static boolean getComparisonValue(ComparisonExpression expression,
+      SalusEnrichedMetric metric, Map<String, Double>customMetricData) {
+    if (expression.getComparator() == Comparator.REGEX_MATCH ||
+        expression.getComparator() == Comparator.NOT_REGEX_MATCH) {
```
This works for now but checking if the keys of `stringCompareFunctions` contain `getComparator` could be more future proof.

itzg commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+    Map.Entry<String, Double> eval(MetricExpressionBase node) {
+      Double part = null, total = null;
+      double percentage;
+      PercentageEvalNode percentageEvalNode = (PercentageEvalNode)node;
```
Does this unconditional casting work because it's the only custom evaluation supported so far?

GeorgeJahad replied:
yes, a shortcut until I get rate/other custom metrics working.

itzg commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+      PercentageEvalNode percentageEvalNode = (PercentageEvalNode)node;
+      // get the partial value
+      for (Metric metric : metrics) {
+        if (metric.getName().equals(percentageEvalNode.getPart())) {
+          if (metric.getValueCase() == ValueCase.INT) {
+            part = (double) metric.getInt();
+          } else {
+            part = metric.getFloat();
+          }
+        }
```
Since this is the second or third time to see this "get double value of metric with given name", a helper method would improve readability.

GeorgeJahad replied:
will do

itzg commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+        // calculate the percent
+        if (part != null && total != null) {
+          percentage = part / total * 100;
+          return new AbstractMap.SimpleEntry<>(percentageEvalNode.getAs(), percentage);
```
In Java 9 they finally added [`Map.entry(k,v)`](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/util/Map.html#entry(K,V)) for this scenario
```suggestion
          return Map.entry(percentageEvalNode.getAs(), percentage);
```

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+          metric.getMetrics());
+      for (MetricExpressionBase customMetric : customMetrics) {
+        Map.Entry<String, Double> entry = evaluator.eval(customMetric);
+        customMetricData.put(entry.getKey(), entry.getValue());
```
I was thinking custom metrics would end up being collected as a group of normal metric objects and then added into `metric.getMetrics()`.  That way we don't have to do anything special with them later on.

GeorgeJahad replied:
@jjbuchan  Yes, adding them to the metric's Metrics list is a good idea.   I'll do it.  I don't understand your final sentence though.  Please clarify:

> Maybe we need to do use methods, so it can be treated like normal in most places here, but then we'd still know which ones are custom to repost back to UMB.

jjbuchan replied:
> Maybe we need to use both methods,

That is what I meant to say.

So we'd keep a list of custom metrics to later send to UMB but also add those into the main metrics list for the other processing in this method.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+      StateEvaluator.MetricEvaluator evaluator = new StateEvaluator.MetricEvaluator(
+          metric.getMetrics());
+      for (MetricExpressionBase customMetric : customMetrics) {
+        Map.Entry<String, Double> entry = evaluator.eval(customMetric);
```
This appears to eliminate the possibility of a custom metric being used to generate another custom metric.

I'm sure this isn't a problem now but I just wanted to make it clear that that's the case.  In kapacitor you could chain custom metrics so that the first one generated was also used to create the second one and so on.

itzg commented on src/main/java/com/rackspace/salus/event/processor/services/UniversalMetricHandler.java:
```diff
@@ -108,12 +108,14 @@ SalusEnrichedMetric convertUniversalMetricToSalusMetric(UniversalMetricFrame uni
     }
-    log.info("TODO Remove tasks from {}", tasksToUndeploy);
+    for (SalusEventEngineTask salusEventEngineTask : tasksToUndeploy) {
+      esperEngine.removeTask(salusEventEngineTask);
```
Can removeTask not work with the superclass `EventEngineTask`? Otherwise, it seems like this code would eventually need to be conditionalized for non-Salus tasks.

jjbuchan replied:
I was wondering the same

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+    }
+    if (operand1 == null) {
+      operand1 = customMetricData.get(expression.getValueName());
+    }
```
Per my comment earlier about custom metrics, this step wouldnt be needed if we treated them like any other metric.

itzg reviewed (commented):
I haven't reviewed unit tests yet, but wanted to submit comments so far.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+      operand1 = customMetricData.get(expression.getValueName());
+    }
+    if (operand1 == null) {
+      throw new IllegalArgumentException("missing/incorrect type for operand in Metrics: "
```
Should we be catching this exception somewhere and logging a failure?

jjbuchan replied:
There might be a few other places like this.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+  // Generate the synthetic metrics, (currently just percent)
+  static class MetricEvaluator {
+    @NonNull
+    private List<Metric> metrics;
```
Maybe this should accept a map of name -> metric instead of a list to avoid potentially iterating over the list many times?

jjbuchan replied:
Nevermind, that's probably an unnecessary optimization at this stage.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+  private static Boolean matchRegex(String s, String pattern) {
+    Pattern p = Pattern.compile(pattern);
+    Matcher m = p.matcher(s);
+    return m.matches();
```
I can look more into that

itzg replied:
Oh yeah, ele might be doing the equivalent of a `find` rather than `matches`

jjbuchan replied:
> The matches method attempts to match the entire input sequence against the pattern.
> The find method scans the input sequence looking for the next subsequence that matches the pattern.

I think `find` is what we want.

jjbuchan reviewed (commented):
Following Geoff's lead and posting initial commentts.

GeorgeJahad commented on src/main/java/com/rackspace/salus/event/processor/engine/EsperEngine.java:
```diff
@@ -171,4 +173,37 @@ public void sendMetric(EnrichedMetric metric) {
+        "select StateEvaluator.generateEnrichedMetric(metric, '%s') " +
+        "from SalusEnrichedMetric(" +
+        "    monitoringSystem='SALUS' and\n" +
+        "    tenantId='%s' and %s) metric;";
```
will do.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+             Comparator.GREATER_THAN, (op1, op2)->(op1 > op2),
+             Comparator.GREATER_THAN_OR_EQUAL_TO, (op1, op2)->(op1 >= op2));
+
+  private static Boolean matchRegex(String s, String pattern) {
```
Also forgot to mention that this method prevents us from using the match found later on.  i.e. in a notiication

e.g. in a match for `4\d\d` it could be good to know if it found `400` vs. `404`.

Might be too much effort to allow for that now, but just wanted to point it out in case others have opinions.

GeorgeJahad replied:
@jjbuchan  I don't quite see how I should do that?  How would the alternative be implemented?

jjbuchan replied:
That's why I said "Might be too much effort to allow for that now."  I'm sure it'd require much more major changes

GeorgeJahad commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -17,33 +17,252 @@
+      }
+    }
+    // sort state expressions
+    Map<TaskState, StateExpression> stateExpressions = new HashMap<>();
+    for (StateExpression stateExpression : data.getEventEngineTask().getTaskParameters()
+        .getStateExpressions()) {
+      stateExpressions.put(stateExpression.getState(), stateExpression);
+    }
+
+    // Check each type of stateExpression and return the first one that is true
+    for (String state : List.of("CRITICAL", "WARNING", "OK")) {
```
i'll try that

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/engine/EsperEngine.java:
```diff
@@ -171,4 +173,37 @@ public void sendMetric(EnrichedMetric metric) {
+        "select StateEvaluator.generateEnrichedMetric(metric, '%s') " +
+        "from SalusEnrichedMetric(" +
+        "    monitoringSystem='SALUS' and\n" +
+        "    tenantId='%s' and %s) metric;";
```
monitorType and monitorScope are missing from this?

jjbuchan replied:
I forgot we still need to add `excludedResourceIds` to Tasks to.  I'll create a jira for that.

GeorgeJahad replied:
yeh, i mentioned it in the todos

GeorgeJahad commented on src/main/java/com/rackspace/salus/event/processor/model/EsperTaskData.java:
```diff
@@ -0,0 +1,28 @@
+  @NonNull
+  String deploymentId;
+  @NonNull
+  EventEngineTask eventEngineTask;
```
i'll look into that, but may have to make it a separate pr for next sprint.

GeorgeJahad commented on src/test/java/com/rackspace/salus/event/processor/services/UniversalMetricListenerTest.java:
```diff
@@ -144,7 +144,6 @@ public void testInitialStartup_2consumers() {
     assertThat(consumer2.getCurrentAssignedPartitions()).isEqualTo(consumer2.getTrackedPartitions());
   }
 
-  @Test
```
typo.  i'll fix.  (This is one of the intermittent failures.)

jjbuchan replied:
I think that means you have a slow laptop.  I would get failures locally until I settled on this value

```
    // increase the delay to help slow laptops running tests locally
    testTimeout = appProperties.getPartitionAssignmentDelay().plus(Duration.ofSeconds(3)).toMillis();
```

jjbuchan commented:
Looks like I ignored your description when I started reading this last night.

> A number of the UniversalMetricHandler tests seemed to be failing intermittently even before I merged my changes. We'll need to investigate those.

There is a timeout field that can be customized.  The value set is what works for me but if your laptop is running slower that would explain why things are failing.  My hope was if it worked for me it would never be too slow and fail in google.

```
    // increase the delay to help slow laptops running tests locally
    testTimeout = appProperties.getPartitionAssignmentDelay().plus(Duration.ofSeconds(3)).toMillis();
```

> Should the event engine really whether the monitorScope is remote or local?

If we are defining monitors remote/local then it makes sense to me that you may have different tasks for remove vs. local?  Without having that field in the task we'd have to rely on customers setting labelSelectors correctly.

It would give them the option to have a single task for all their SQL monitors whether local or remote, but if they wanted to have different criteria for each they would have to know to set it in the label selector.

I definitely see the benefit of not having it, but I think I prefer the consistency between monitor and task configurations.

> Should this PR use the message field in state expression? It is not currently.

This PR doesn't need to.  We can have a jira issue specifically to handle that everywhere.

jjbuchan commented on src/test/java/com/rackspace/salus/event/processor/services/StateEvaluatorTest.java:
```diff
@@ -0,0 +1,290 @@
+  }
+
+  @Test
+  public void badParameterTest() {
```
Using JUnit's nested tests might be better than grouping all the assertions together like this?

jjbuchan reviewed (commented):
Tests looks good but would be nice to include some for epl generation and `removeTask`.

itzg reviewed (commented):
All files reviewed now

GeorgeJahad commented:
Ok @jjbuchan and @itzg , I think I've addressed all the comments except this one:

stream().sorted() comment Geoff made here: https://github.com/racker/salus-event-engine-processor/pull/4/files#r512812264

and I haven't added any metrics just log messages.

itzg reviewed (approved):
The `stream()` suggestion was totally optional, so I'm good with what you have now.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/engine/EsperEngine.java:
```diff
@@ -174,25 +178,38 @@ public void sendMetric(EnrichedMetric metric) {
+    String eplString = createTaskEpl(t);
     EPStatement epStatement = compileAndDeployQuery(eplString);
     StateEvaluator.saveTaskData(taskId, epStatement.getDeploymentId(), t);
+    log.info("Adding task epl: {}", eplString);
```
Should this be debug or trace instead?

I think an info message should contain things like taskId and tenantId rather than the epl.

jjbuchan replied:
and I think we've decided on using the format like `task={}` in log messages for assigning values to a key.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/engine/EsperEngine.java:
```diff
@@ -201,8 +218,9 @@ public void removeTask(EventEngineTask t) {
     if (deploymentId != null) {
       try {
         runtime.getDeploymentService().undeploy(deploymentId);
+        log.info("Removed task: {}", taskId);
```
Include tenantId as well?

jjbuchan replied:
```suggestion
        log.info("Removing task={} for tenant={}", taskId, t.getTenantId());
```

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/engine/EsperEngine.java:
```diff
@@ -201,8 +218,9 @@ public void removeTask(EventEngineTask t) {
+        log.info("Removed task: {}", taskId);
       } catch (EPUndeployException e) {
-        e.printStackTrace();
+        log.warn("Execption removing task: {}", taskId);
```
```suggestion
        log.warn("Exception removing task={} for tenant={}", taskId, t.getTenantId());
```

jjbuchan replied:
Could include the message from `e` too

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/services/StateEvaluator.java:
```diff
@@ -67,13 +70,15 @@ public static void saveTaskData(String taskId, String deploymentId, EventEngineT
-    if (!taskDataMap.containsKey(taskId)) {
+    EsperTaskData taskData = taskDataMap.remove(taskId);
+    if (taskData == null) {
+      log.warn("Could not remove task data for: {}", taskId);
```
Following on from my other comments, updating all these log messages to be of the `key={}` format would be good.

jjbuchan reviewed (commented):
Looks good.  Just some logging comments.

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/engine/EsperEngine.java:
```diff
@@ -171,4 +177,52 @@ public void sendMetric(EnrichedMetric metric) {
+    String eplString = createTaskEpl(t);
+    EPStatement epStatement = compileAndDeployQuery(eplString);
+    StateEvaluator.saveTaskData(taskId, epStatement.getDeploymentId(), t);
+    log.trace("Adding task for tenant={}, task={}", t.getTenantId(), taskId);
```
I can't be sure since our logging format isn't finalized but I do wonder if the `,` directly after `{}` will lead to it being included in the value string.

jjbuchan replied:
@itzg do you know if this'll be a problem or not?

GeorgeJahad replied:
i'll just remove them for now.  not worth thinking about

jjbuchan commented on src/main/java/com/rackspace/salus/event/processor/engine/EsperEngine.java:
```diff
@@ -171,4 +177,52 @@ public void sendMetric(EnrichedMetric metric) {
+        runtime.getDeploymentService().undeploy(deploymentId);
+        log.trace("Removing task for tenant={} task={}", t.getTenantId(), taskId);
+      } catch (EPUndeployException e) {
+        log.trace("Exception removing task for tenant={} task={}, exception message={}",
```
```suggestion
        log.trace("Exception removing task for tenant={} task={} exception message={}",
```

jjbuchan reviewed (commented):
last one

jjbuchan reviewed (approved):
I see this is still `+740 −21` so I'm assuming all is as it was before.
